File: rx-codelab/src/test/java/com/jraska/rx/codelab/solution/Solution_Task2_Transformations.java
package com.jraska.rx.codelab.solution;

import com.jraska.rx.codelab.RxLogging;
import com.jraska.rx.codelab.http.GitHubApi;
import com.jraska.rx.codelab.http.GitHubConverter;
import com.jraska.rx.codelab.http.HttpModule;
import com.jraska.rx.codelab.http.User;
import io.reactivex.Observable;
import org.junit.Before;
import org.junit.Test;

public class Solution_Task2_Transformations {
  private static final String LOGIN = "defunkt";
  private GitHubApi gitHubApi = HttpModule.mockedGitHubApi();

  @Before
  public void setUp() {
    RxLogging.enableObservableSubscribeLogging();
  }

  @Test
  public void map_convertUserDto() {
    gitHubApi.getUser(LOGIN)
      .map(GitHubConverter::convert)
      .map(User::toString)
      .subscribe(System.out::println);
  }

  @Test
  public void flatMap_getFirstUserDetailAfterGettingList() {
    gitHubApi.getFirstUsers()
      .flatMap(gitHubUsers -> gitHubApi.getUser(gitHubUsers.get(0).login))
      .map(GitHubConverter::convert)
      .subscribe(System.out::println);
  }

  @Test
  public void replayAutoConnect_oneRequestForTwoSubscriptions() {
    Observable<User> observable = gitHubApi.getUser(LOGIN)
      .map(GitHubConverter::convert)
      .replay(1)
      .autoConnect();

    observable.subscribe(System.out::println);
    observable.subscribe(System.out::println);
  }
}

File: rx-codelab/src/test/java/com/jraska/rx/codelab/solution/SolutionTask2Transformations.kt
package com.jraska.rx.codelab.solution

import com.jraska.rx.codelab.RxLogging
import com.jraska.rx.codelab.http.GitHubConverter
import com.jraska.rx.codelab.http.HttpModule
import org.junit.Before
import org.junit.Test

class SolutionTask2Transformations {
  private val gitHubApi = HttpModule.mockedGitHubApi()

  @Before
  fun setUp() {
    RxLogging.enableObservableSubscribeLogging()
  }

  @Test
  fun map_convertUserDto() {
    gitHubApi.getUser(LOGIN)
      .map { GitHubConverter.convert(it) }
      .map { it.toString() }
      .subscribe { println(it) }
  }

  @Test
  fun flatMap_getFirstUserDetailAfterGettingList() {
    gitHubApi.getFirstUsers()
      .flatMap { gitHubUsers -> gitHubApi.getUser(gitHubUsers[0].login) }
      .map { GitHubConverter.convert(it) }
      .subscribe { println(it) }
  }

  @Test
  fun replayAutoConnect_oneRequestForTwoSubscriptions() {
    val observable = gitHubApi.getUser(LOGIN)
      .map { GitHubConverter.convert(it) }
      .replay(1)
      .autoConnect()

    observable.subscribe { println(it) }
    observable.subscribe { println(it) }
  }

  companion object {
    private const val LOGIN = "defunkt"
  }
}
